### PR TITLE
LEAF-3776: AD Distro Group Integration Part 1

### PR DESCRIPTION
**Pull request by raphaelfretter**
The only functional change is the save button in import groups says 'import'. AD API endpoints are inaccessible but in place until AD server routing issue is resolved.

### CODE REVIEW

File: LEAF_Nexus/sources/ActiveDirectory.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+      $this->bind = ldap_bind($this->conn, $this->rdn, AD_PW);
+   }
+
+   public function searchMember(string $uid = "") : ?array
+   {
+      $info = [];
+      $filter = "(&(objectCategory=Person)(objectClass=User)(anr=$uid))"; // search only for users classified as persons (i.e. not service accounts)
+      $attr = ["objectClass", "sn", "givenName","initials","title","description","telephoneNumber","mail","sAMAccountName","objectGUID","mobile","physicalDeliveryOfficeName"]; // fields for each result to return
+      if ($this->bind) {
```
Comment (Pelentan) [inline LEAF_Nexus/sources/ActiveDirectory.php:32]: This isn't a boolean variable.

File: LEAF_Nexus/sources/ActiveDirectory.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+         $results = ldap_search($this->conn, $this->base, $filter, $attr);
+         $info = ldap_get_entries($this->conn, $results);
+      }
+      return $info;
```
Comment (Pelentan) [inline LEAF_Nexus/sources/ActiveDirectory.php:36]: Does this need to be in our new format?  I.e. returning a json-style with results either being 1 or 2 and that data being returned in another field?  If so, please ask Jamie for the exact field names.

File: LEAF_Nexus/sources/ActiveDirectory.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+   public function searchGroup(string $input = "") : array
+   {
+      $info = [];
+      $filter = "(&(objectClass=Group)(anr=$input))"; // search only for groups
+      $attr = ["objectClass", "cn", "title", "sAMAccountName", "managedBy", "member", "description", "objectGUID"]; // fields for each result to return
+      if ($this->bind) {
+         $results = ldap_search($this->conn, $this->base, $filter, $attr);
+         $info = ldap_get_entries($this->conn, $results);
+      }
+      return $info;
```
Comment (Pelentan) [inline LEAF_Nexus/sources/ActiveDirectory.php:48]: same.

File: LEAF_Nexus/sources/ActiveDirectory.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+   {
+      $info = [];
+      $fmtMembers = implode(")(", $members);
+      $filter = "(/($fmtMembers))"; /// search only for users classified as persons (i.e. not service accounts)
+      $attr = ["objectClass", "sn", "givenName","initials","title","description","telephoneNumber","mail","sAMAccountName","objectGUID","mobile","physicalDeliveryOfficeName"]; // fields for each result to return
+      if ($this->bind) {
+         $results = ldap_search($this->conn, $this->base, $filter, $attr);
+         $info = ldap_get_entries($this->conn, $results);
+      }
+      return $info;
```
Comment (Pelentan) [inline LEAF_Nexus/sources/ActiveDirectory.php:61]: ditto

**Review by Pelentan — CHANGES_REQUESTED** [review]

**Review by Pelentan — APPROVED** [review]

File: LEAF_Nexus/sources/ActiveDirectory.php [hunk 5]
```diff
@@ -29,23 +29,53 @@ public function searchMember(string $uid = "") : ?array
...
+         $return_value = array(
+            "status" => array(
+               "code" => 4,
+               "message" => "The connection to the LDAP server failed."
+            )
+         );
+         return $return_value;
       }
-      return $info;
+
```
Comment (jampaul3) [inline LEAF_Nexus/sources/ActiveDirectory.php:41]: It's better to have an else here and only one return for this method. If you have an else here you won't need the return on line 39 you can wait and only have the one return on line 51

File: LEAF_Nexus/sources/ActiveDirectory.php [hunk 6]
```diff
@@ -29,23 +29,53 @@ public function searchMember(string $uid = "") : ?array
...
-         $info = ldap_get_entries($this->conn, $results);
+      if ($this->bind === false) {
+         $return_value = array(
+            "status" => array(
+               "code" => 4,
+               "message" => "The connection to the LDAP server failed."
+            )
+         );
+         return $return_value;
       }
```
Comment (jampaul3) [inline LEAF_Nexus/sources/ActiveDirectory.php:67]: Same

File: LEAF_Nexus/sources/ActiveDirectory.php [hunk 7]
```diff
@@ -54,10 +84,25 @@ public function listMembers(array $members = []) : array
...
```
Comment (jampaul3) [inline LEAF_Nexus/sources/ActiveDirectory.php:95]: And again

**Review by jampaul3 — CHANGES_REQUESTED** [review]

**Review by aerinkayne — APPROVED** [review]

**Review by shaneodd — APPROVED** [review]

**Pull request merged**